mypy/semanal_shared.py

```diff
@@ -378,7 +379,24 @@ def find_dataclass_transform_spec(node: Node | None) -> DataclassTransformSpec |
         # `@dataclass_transform(...)` syntax and never `@dataclass_transform`
         node = node.func
 
+    # For functions, we can directly consult the AST field for the spec
     if isinstance(node, FuncDef):
         return node.dataclass_transform_spec
 
+    if isinstance(node, ClassDef):
+        node = node.info
+    if isinstance(node, TypeInfo):
+        # Search all parent classes to see if any are decorated with `typing.dataclass_transform`
+        for base in node.mro[1:]:
+            if base.defn.dataclass_transform_spec is not None:
+                return base.defn.dataclass_transform_spec
+
+        # Check if there is a metaclass that is decorated with `typing.dataclass_transform`
+        metaclass_type = node.metaclass_type
+        if (
+            metaclass_type is not None
+            and metaclass_type.type.defn.dataclass_transform_spec is not None
```

Comment:
This is kind of odd; from [the PEP](https://peps.python.org/pep-0681/#the-dataclass-transform-decorator):

> If `dataclass_transform` is applied to a class, dataclass-like semantics will be assumed for any class that directly or indirectly derives from the decorated class or uses the decorated class as a metaclass. Attributes on the decorated class and its base classes are not considered to be fields.

Anything that inherits from a class with `dataclass_transform_spec` must have dataclass-like semantics. I don't think the _precludes_ it from being used a metaclass, but that would seem kind of unnatural? The PEP also uses the wording "uses the decorated class as a metaclass", and I'm not sure if that's semantically equivalent to "uses a class inheriting from the decorated class as a metaclass" (I would assume it only means the directly-specified metaclass).

That said, Pyright _does_ seem to support it, and the PEP names Pyright as the reference implementation.

Reply:
I think I was misunderstanding Pyright's behavior earlier; it didn't report an error for this example:

```
from typing import dataclass_transform

@dataclass_transform()
class BaseMeta(type): ...

class SubMeta(BaseMeta): ...

class Foo(metaclass=SubMeta):
    foo: int
Foo(foo=0)
```

Pyright doesn't detect any errors here, but that's because it's abandoning analysis due to an arbitrary metaclass rather than due to detecting that it's a proper dataclass transform. For example, it also accepts invalid expressions such as `Foo(foo=1, bar=2)`. If `BaseMeta` is used as the metaclass instead, it detects these errors correctly.

This seems to match my interpretation above, and leads me to believe that we should only support the immediate metaclass and not any of its parent classes.